### PR TITLE
Issue-1394  pebble support 

### PR DESCRIPTION
**Pull request by tonyqzhang**
## Why this should be merged
to support pebble

### CODE REVIEW

File: database/pebble/db.go [hunk 1]
```diff
@@ -0,0 +1,458 @@
...
+
+	return &iter{
+		db:   db,
+		iter: db.db.NewIter(&pebble.IterOptions{LowerBound: start}),
+	}
+}
+
+// bytesPrefix returns key range that satisfy the given prefix.
+// This only applicable for the standard 'bytes comparer'.
+func bytesPrefix(prefix []byte) *pebble.IterOptions {
```
Comment (danlaine) [inline database/pebble/db.go:345]: Can we add unit tests for this?

File: database/pebble/db.go [hunk 2]
```diff
@@ -0,0 +1,458 @@
...
+func (it *iter) Value() []byte {
+	if !it.valid {
+		return nil
+	}
+	return slices.Clone(it.iter.Value())
+}
+
+func (it *iter) Release() {
+	db := it.db
+	if _, herr := db.HealthCheck(context.TODO()); herr != nil {
```
Comment (danlaine) [inline database/pebble/db.go:440]:
```suggestion
	if _, herr := it.db.HealthCheck(context.TODO()); herr != nil {
```

File: database/pebble/db_test.go [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+)
+
+func TestInterface(t *testing.T) {
+	for _, test := range database.Tests {
+		folder := t.TempDir()
+		cfg := NewDefaultConfig()
+		db, err := New(folder, cfg, logging.NoLog{}, "pebble", prometheus.NewRegistry())
+		if err != nil {
+			t.Fatalf("pebble.New(%q, logging.NoLog{}) errored with %s", folder, err)
+		}
```
Comment (danlaine) [inline database/pebble/db_test.go:23]:
```suggestion
		db, err := New(folder, cfg, logging.NoLog{}, "pebble", prometheus.NewRegistry())
		require.NoError(t, err)
```

File: database/pebble/db_test.go [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+}
+
+func FuzzInterface(f *testing.F) {
+	for _, test := range database.FuzzTests {
+		folder := f.TempDir()
+		cfg := NewDefaultConfig()
+		db, err := New(folder, cfg, logging.NoLog{}, "", prometheus.NewRegistry())
+		if err != nil {
+			require.NoError(f, err)
+		}
```
Comment (danlaine) [inline database/pebble/db_test.go:42]:
```suggestion
		db, err := New(folder, cfg, logging.NoLog{}, "", prometheus.NewRegistry())
		require.NoError(f, err)
```

**Review by danlaine — CHANGES_REQUESTED** [review]

File: database/pebble/db.go [hunk 5]
```diff
@@ -0,0 +1,458 @@
...
+	// closeCh is closed when Close() is called.
+	closeCh chan struct{}
+}
+
+type Config struct {
+	CacheSize                   int // B
+	BytesPerSync                int // B
+	WALBytesPerSync             int // B (0 disables)
+	MemTableStopWritesThreshold int // num tables
+	MemTableSize                int // B
```
[the remark this answers is not in the document]
Comment (tonyqzhang) [reply database/pebble/db.go:46]: It means Byte. It's updated to Byte

File: database/pebble/db.go [hunk 6]
```diff
@@ -0,0 +1,458 @@
...
+	_ database.Iterator = (*iter)(nil)
+
+	ErrInvalidOperation = errors.New("invalid operation")
+)
+
+type Database struct {
+	db *pebble.DB
+
+	closed utils.Atomic[bool]
+	// closeOnce sync.Once
```
[the remark this answers is not in the document]
Comment (tonyqzhang) [reply database/pebble/db.go:36]: Absolutely agree! Removed

File: database/pebble/db.go [hunk 7]
```diff
@@ -0,0 +1,458 @@
...
+}
+
+func NewDefaultConfig() Config {
+	return Config{
+		CacheSize:                   1024 * 1024 * 1024,
+		BytesPerSync:                1024 * 1024,
+		WALBytesPerSync:             1024 * 1024,
+		MemTableStopWritesThreshold: 8,
+		MemTableSize:                16 * 1024 * 1024,
+		MaxOpenFiles:                4_096,
```
[the remark this answers is not in the document]
Comment (tonyqzhang) [reply database/pebble/db.go:57]: Agree! updated it

File: database/pebble/db.go [hunk 8]
```diff
@@ -0,0 +1,458 @@
...
+	// TODO: add metrics support
+	return wrappedDB, nil
+}
+
+func (db *Database) Close() error {
+	// close a db twice will trigger panic by pebble instead of error
+	if _, herr := db.HealthCheck(context.TODO()); herr != nil {
+		return herr
+	}
+	db.closed.Set(true)
```
[the remark this answers is not in the document]
Comment (tonyqzhang) [reply database/pebble/db.go:114]: I took the former. Leave the HealthCheck function for other purpose later

File: database/pebble/db.go [hunk 9]
```diff
@@ -0,0 +1,458 @@
...
+	if _, herr := db.HealthCheck(context.TODO()); herr != nil {
+		return false, herr
+	}
+
+	_, closer, err := db.db.Get(key)
+	if err == pebble.ErrNotFound {
+		return false, nil
+	} else if err != nil {
+		return false, updateError(err)
+	}
```
[the remark this answers is not in the document]
Comment (tonyqzhang) [reply database/pebble/db.go:145]: OK, changed it!

File: database/pebble/db.go [hunk 10]
```diff
@@ -0,0 +1,458 @@
...
+	return true, closer.Close()
+}
+
+// Get returns the value the key maps to in the database
+func (db *Database) Get(key []byte) ([]byte, error) {
+	if _, herr := db.HealthCheck(context.TODO()); herr != nil {
+		return nil, herr
+	}
```
[the remark this answers is not in the document]
Comment (tonyqzhang) [reply database/pebble/db.go:153]: use  db.closed.Get() to replace HealthCheck

File: database/pebble/db.go [hunk 11]
```diff
@@ -0,0 +1,458 @@
...
+	db    *Database
+	size  int
+
+	// Support BATCH_REWRITE
+	applied atomic.Bool
+}
+
+// NewBatch creates a write/delete-only buffer that is atomically committed to
+// the database when write is called
+func (db *Database) NewBatch() database.Batch { return &batch{db: db, batch: db.db.NewBatch()} }
```
[the remark this answers is not in the document]
Comment (tonyqzhang) [reply database/pebble/db.go:228]: Ok, changed it

File: database/pebble/db.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (danlaine) [reply database/pebble/db.go:36]: Did you forget to push? Looks like there are no new commits

File: database/pebble/db.go (hunk 2 above)
Comment (tonyqzhang) [reply database/pebble/db.go:440]: replace by it.db.closed.Get()

File: database/pebble/db_test.go [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+	for _, size := range database.BenchmarkSizes {
+		keys, values := database.SetupBenchmark(b, size[0], size[1], size[2])
+		for _, bench := range database.Benchmarks {
+			folder := b.TempDir()
+			cfg := NewDefaultConfig()
+
+			db, err := New(folder, cfg, logging.NoLog{}, "", prometheus.NewRegistry())
+			if err != nil {
+				b.Fatal(err)
+			}
```
[the remark this answers is not in the document]
Comment (tonyqzhang) [reply database/pebble/db_test.go:64]: Agree! Changed it

File: database/pebble/db.go [hunk 13]
```diff
@@ -0,0 +1,458 @@
...
+}
+
+func (db *Database) HealthCheck(_ context.Context) (interface{}, error) {
+	if db.closed.Get() {
+		return nil, database.ErrClosed
+	}
+	return nil, nil
+}
+
+// Has returns if the key is set in the database
```
[the remark this answers is not in the document]
Comment (tonyqzhang) [reply database/pebble/db.go:134]: Agree! Changed it

File: database/pebble/db_test.go (hunk 3 above)
Comment (tonyqzhang) [reply database/pebble/db_test.go:23]: Changed it!

File: database/pebble/db_test.go (hunk 4 above)
Comment (tonyqzhang) [reply database/pebble/db_test.go:42]: Changed it

File: database/pebble/db.go [hunk 14]
```diff
@@ -0,0 +1,458 @@
...
+// Size retrieves the amount of data queued up for writing.
+func (b *batch) Size() int { return b.size }
+
+// Write flushes any accumulated data to disk.
+func (b *batch) Write() error {
+	if _, herr := b.db.HealthCheck(context.TODO()); herr != nil {
+		return herr
+	}
+
+	// Support BATCH_REWRITE
```
[the remark this answers is not in the document]
Comment (tonyqzhang) [reply database/pebble/db.go:251]: changed it!

File: database/pebble/db.go [hunk 15]
```diff
@@ -0,0 +1,458 @@
...
+	if _, herr := db.HealthCheck(context.TODO()); herr != nil {
+		return herr
+	}
+
+	err := updateError(db.db.Compact(start, limit, true))
+	if err != nil {
+		// The default compare function of pebble is bytes.Compare.
+		// Use this default compare function since we don't setup
+		// the compare function when we create pebble db.
+		// It's imporssible to retrieve this default
```
[the remark this answers is not in the document]
Comment (tonyqzhang) [reply database/pebble/db.go:200]: Thanks! Changed it!

File: database/pebble/db.go [hunk 16]
```diff
@@ -0,0 +1,458 @@
...
+// Write flushes any accumulated data to disk.
+func (b *batch) Write() error {
+	if _, herr := b.db.HealthCheck(context.TODO()); herr != nil {
+		return herr
+	}
+
+	// Support BATCH_REWRITE
+	// the underline pebble db doesn't support batch rewrite but got panic instead
+	// we have to create a new batch which is a kind of duplicate of the given
+	// batch(arg b) and commit this new batch on behalf of the given batch
```
[the remark this answers is not in the document]
Comment (tonyqzhang) [reply database/pebble/db.go:254]: Thanks! Changed it

**Comment by tonyqzhang** [comment]
Not yet! Will push soon. 
    On Tuesday, May 16, 2023 at 01:45:46 PM PDT, Dan Laine ***@***.***> wrote:  
 
 


@danlaine commented on this pull request.

In database/pebble/db.go:
> +	"golang.org/x/exp/slices"
+)
+
+var (
+	_ database.Database = (*Database)(nil)
+	_ database.Batch    = (*batch)(nil)
+	_ database.Iterator = (*iter)(nil)
+
+	ErrInvalidOperation = errors.New("invalid operation")
+)
+
+type Database struct {
+	db *pebble.DB
+
+	closed utils.Atomic[bool]
+	// closeOnce sync.Once

Did you forget to push? Looks like there are no new commits

**Comment by tonyqzhang** [comment]
Just pushed for all resolved issues.Thanks!Tony Zhang
    On Tuesday, May 16, 2023 at 01:45:46 PM PDT, Dan Laine ***@***.***> wrote:  
 
 


@danlaine commented on this pull request.

In database/pebble/db.go:
> +	"golang.org/x/exp/slices"
+)
+
+var (
+	_ database.Database = (*Database)(nil)
+	_ database.Batch    = (*batch)(nil)
+	_ database.Iterator = (*iter)(nil)
+
+	ErrInvalidOperation = errors.New("invalid operation")
+)
+
+type Database struct {
+	db *pebble.DB
+
+	closed utils.Atomic[bool]
+	// closeOnce sync.Once

Did you forget to push? Looks like there are no new commits

File: database/pebble/db.go (hunk 1 above)
Comment (tonyqzhang) [reply database/pebble/db.go:344]: added and pushed

File: database/pebble/db.go [hunk 17]
```diff
@@ -0,0 +1,456 @@
...
+// pebble returns error but we like to return nil
+func (db *Database) Compact(start []byte, limit []byte) error {
+	// Compact causes panic if the db has already been closed
+	if db.closed.Get() {
+		return database.ErrClosed
+	}
+
+	err := updateError(db.db.Compact(start, limit, true))
+	if err != nil && bytes.Equal(start, limit) {
+		// The default compare function of pebble is bytes.Compare.
```
Comment (hexfusion) [inline database/pebble/db.go:191]: Just to confirm the original compare was  `bytes.Compare(...) == 0` do you have a ref to the original check?

File: database/pebble/db.go [hunk 18]
```diff
@@ -0,0 +1,458 @@
...
+		// Don't Commit it again, got panic otherwise
+		// Create a new batch to do Commit
+		newbatch := &batch{db: b.db, batch: b.db.db.NewBatch()}
+
+		// duplicate b.batch to newbatch.batch
+		if err := newbatch.batch.Apply(b.batch, nil); err != nil {
+			return err
+		}
+		return updateError(newbatch.batch.Commit(pebble.NoSync))
+	}
```
[the remark this answers is not in the document]
Comment (tonyqzhang) [reply database/pebble/db.go:265]: test_database.go defines TestBatchRewrite
// TestBatchRewrite tests to make sure that write can be called multiple times
// on a batch and the values will be updated correctly.

File: database/pebble/db.go [hunk 19]
```diff
@@ -0,0 +1,458 @@
...
+		// compare function from the pebble package since
+		// it's defined in internal.
+		// We use bytes.Equal instead of bytes.Compare
+		// because golangci_lint recommends
+		if bytes.Equal(start, limit) {
+			// Don't return error if it is caused
+			// by the same of start and limit
+			// Do nothing just ignore the error
+			// since there is no need to compact
+			err = nil
```
[the remark this answers is not in the document]
Comment (tonyqzhang) [reply database/pebble/db.go]: test_database.go defines TestCompactNoPanic: require.NoError(db.Compact(nil, nil))
I think it's just fine if there is nothing to compact.

File: database/pebble/db.go [hunk 20]
```diff
@@ -0,0 +1,458 @@
...
+	if _, herr := db.HealthCheck(context.TODO()); herr != nil {
+		return nil, herr
+	}
+
+	data, closer, err := db.db.Get(key)
+	if err != nil {
+		return nil, updateError(err)
+	}
+	ret := make([]byte, len(data))
+	copy(ret, data)
```
[the remark this answers is not in the document]
Comment (tonyqzhang) [reply database/pebble/db.go:154]: Great! changed it

File: database/pebble/db.go [hunk 21]
```diff
@@ -0,0 +1,458 @@
...
+	}
+	db.closed.Set(true)
+
+	err := updateError(db.db.Close())
+	if err != nil {
+		if strings.Contains(err.Error(), "leaked iterator") {
+			// avalanche database support close db w/o error
+			// even if there is an iterator which is not released
+			return nil
+		}
```
[the remark this answers is not in the document]
Comment (tonyqzhang) [reply database/pebble/db.go]: test_database.go defines TestIteratorError:
        // Call Next() and ensure that if the database is closed, the iterator
	// can still report the current contents.
	require.True(iterator.Next())
	require.NoError(db.Close())

I really hate to take advantage of specific string from error. I agree with you. This is really fragile. But I didn't find any way  to detect if there is a "leaked iterator" before closing the db. I'll change this code if I can find a better way to detect "leaked iterator" later

File: database/pebble/db.go [hunk 22]
```diff
@@ -0,0 +1,458 @@
...
+	applied atomic.Bool
+}
+
+// NewBatch creates a write/delete-only buffer that is atomically committed to
+// the database when write is called
+func (db *Database) NewBatch() database.Batch { return &batch{db: db, batch: db.db.NewBatch()} }
+
+// Put the value into the batch for later writing
+func (b *batch) Put(key, value []byte) error {
+	b.size += len(key) + len(value) + 8 // TODO: find byte overhead
```
[the remark this answers is not in the document]
Comment (tonyqzhang) [reply database/pebble/db.go:230]: removed it.

File: database/pebble/db.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (tonyqzhang) [reply database/pebble/db.go]: Just checked it again. Avalanche Database expects Compact(nil, nil) should compact the entire database. I think we miss a compact test case to confirm the compaction really does good job. I'll do more research on this issue.

File: database/pebble/db.go (hunk 22 above)
[the remark this answers is not in the document]
Comment (tonyqzhang) [reply database/pebble/db.go:230]: defined const pebbleByteOverHead for this purpose

File: database/pebble/db.go [hunk 23]
```diff
@@ -0,0 +1,458 @@
...
+		WALBytesPerSync:             cfg.WALBytesPerSync,
+		MemTableStopWritesThreshold: cfg.MemTableStopWritesThreshold,
+		MemTableSize:                cfg.MemTableSize,
+		MaxOpenFiles:                cfg.MaxOpenFiles,
+		MaxConcurrentCompactions:    runtime.NumCPU,
+		Levels:                      make([]pebble.LevelOptions, 7),
+		// TODO: add support for adding a custom logger
+	}
+	// Default configuration sourced from:
+	// https://github.com/cockroachdb/pebble/blob/master/cmd/pebble/db.go#L76-L86
```
[the remark this answers is not in the document]
Comment (tonyqzhang) [reply database/pebble/db.go:78]: changed!

File: database/pebble/db.go [hunk 24]
```diff
@@ -0,0 +1,458 @@
...
+		BytesPerSync:                1024 * 1024,
+		WALBytesPerSync:             1024 * 1024,
+		MemTableStopWritesThreshold: 8,
+		MemTableSize:                16 * 1024 * 1024,
+		MaxOpenFiles:                4_096,
+	}
+}
+
+func New(file string, cfg Config, log logging.Logger, _ string, reg prometheus.Registerer) (database.Database, error) {
+	// These default settings are based on https://github.com/ethereum/go-ethereum/blob/master/ethdb/pebble/pebble.go
```
[the remark this answers is not in the document]
Comment (tonyqzhang) [reply database/pebble/db.go:61]: changed

File: database/pebble/db.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (tonyqzhang) [reply database/pebble/db.go]: Rewrite Compact to fix some conflicts between avalanche DB and pebble
1> avalanche DB treats a nil limit as a key after all keys in the entire DB
     pebble treats a nil limit as a key before all keys in the DB instead
2>  avalanche DB will do compact for the entire DB if both start and limit are nil
     pebble will do nothing and return error
3> If start and limit are the same, not nil, avalanche DB will do nothing and return nil. pebble will do nothing but return error.

File: database/pebble/db.go [hunk 25]
```diff
@@ -0,0 +1,463 @@
...
+		Cache:        pebble.NewCache(int64(cfg.CacheSize)),
+		BytesPerSync: cfg.BytesPerSync,
+		// Although we use `pebble.NoSync`, we still keep the WAL enabled. Pebble
+		// will fsync the WAL during shutdown and should ensure the db is
+		// recoverable if shutdown correctly.
+		WALBytesPerSync:             cfg.WALBytesPerSync,
+		MemTableStopWritesThreshold: cfg.MemTableStopWritesThreshold,
+		MemTableSize:                cfg.MemTableSize,
+		MaxOpenFiles:                cfg.MaxOpenFiles,
+		MaxConcurrentCompactions:    runtime.NumCPU,
```
Comment (patrick-ogrady) [inline database/pebble/db.go:79]: I'm not sure we want this many concurrent compactions. Pebble defaults to 1.

File: database/pebble/db.go [hunk 26]
```diff
@@ -0,0 +1,463 @@
...
+		// compare function from the pebble package since
+		// it's defined in internal.
+		// We use bytes.Equal instead of bytes.Compare
+		// because golangci_lint recommends
+
+		// pebble return error if start and limit are the same,
+		// even if both are nil.
+		// Do nothing just ignore the error if start and limt are the same
+		// and not nil since there is no need to compact and avalanche db
+		// expects a nil return.
```
Comment (joshua-kim) [inline database/pebble/db.go:207]: 1. I feel like this comment is a bit too wordy/confusing, I feel like we should just delete it
2. If this if-check only exists because pebble bubble up an error when you have try to compact where `start == limit` can we just bubble up the error instead of silently failing compaction and remove this case? I think all the callers of this just log an error so i think it should be okay

File: database/pebble/db.go [hunk 27]
```diff
@@ -0,0 +1,463 @@
...
+
+func (b *batch) Size() int { return b.size }
+
+func (b *batch) Write() error {
+	// write causes panic if the db has already been closed
+	if b.db.closed.Get() {
+		return database.ErrClosed
+	}
+
+	// Support batch rewrite
```
Comment (joshua-kim) [inline database/pebble/db.go:261]: Maybe a dumb question but what's the use-case for a batch rewrite?

File: database/pebble/db.go [hunk 28]
```diff
@@ -0,0 +1,463 @@
...
+
+	return &iter{
+		db:   db,
+		iter: db.db.NewIter(&pebble.IterOptions{LowerBound: start}),
+	}
+}
+
+// bytesPrefix returns key range that satisfy the given prefix.
+// This only applicable for the standard 'bytes comparer'.
+func bytesPrefix(prefix []byte) *pebble.IterOptions {
```
Comment (joshua-kim) [inline database/pebble/db.go:353]: I feel like instead of returning `pebble.IterOptions` here we could just return `prefix` and `limit` separately and let the caller create the config.

**Review by joshua-kim — CHANGES_REQUESTED** [review]

File: database/pebble/db_test.go [hunk 29]
```diff
@@ -0,0 +1,91 @@
...
+		cfg := NewDefaultConfig()
+		db, err := New(folder, cfg, logging.NoLog{}, "pebble", prometheus.NewRegistry())
+		require.NoError(t, err)
+		defer db.Close()
+
+		test(t, db)
+
+		// The database may have been closed by the test, so we don't care if it
+		// errors here.
+		_ = db.Close()
```
Comment (joshua-kim) [inline database/pebble/db_test.go:29]: 1. remove this comment
2. can't we just check for `require.NoError` here?

File: database/pebble/db_test.go [hunk 30]
```diff
@@ -0,0 +1,91 @@
...
+	}
+}
+
+func FuzzInterface(f *testing.F) {
+	for _, test := range database.FuzzTests {
+		folder := f.TempDir()
+		cfg := NewDefaultConfig()
+		db, err := New(folder, cfg, logging.NoLog{}, "", prometheus.NewRegistry())
+		require.NoError(f, err)
+		defer db.Close()
```
Comment (joshua-kim) [inline database/pebble/db_test.go:39]: Another duplicate close

File: database/pebble/db_test.go [hunk 31]
```diff
@@ -0,0 +1,91 @@
...
+
+		test(f, db)
+
+		// The database may have been closed by the test, so we don't care if it
+		// errors here.
+		_ = db.Close()
```
Comment (joshua-kim) [inline database/pebble/db_test.go:45]: Another duplicate close

File: database/pebble/db_test.go [hunk 32]
```diff
@@ -0,0 +1,91 @@
...
+			bench(b, db, "pebble", keys, values)
+
+			// The database may have been closed by the test, so we don't care if it
+			// errors here.
+			_ = db.Close()
+		}
+	}
+}
+
+func Test_bytesPrefix(t *testing.T) {
```
Comment (joshua-kim) [inline database/pebble/db_test.go:67]: I feel like instead of testing the `bytesPrefix` method (personally I am generally against testing non-exported code in a package) I think it would make more sense to implicitly test this by doing something like inserting a set of keys into the db and iterating over the range and verifying that the range we got from the iterator is what we expected

File: database/pebble/db.go [hunk 33]
```diff
@@ -0,0 +1,463 @@
...
+	// But pebble treats a nil, no matter start or limit, as a key before
+	// all keys in the DB
+	it := db.db.NewIter(&pebble.IterOptions{})
+	if it.Last() {
+		if lastkey := it.Key(); lastkey != nil {
+			return updateError(db.db.Compact(start, lastkey, true))
+		}
+	}
+
+	return database.ErrNotFoundLastKey
```
Comment (joshua-kim) [inline database/pebble/db.go:223]: Wondering if we can just return nil here

File: database/pebble/db.go [hunk 34]
```diff
@@ -0,0 +1,463 @@
...
+		BytesPerSync: cfg.BytesPerSync,
+		// Although we use `pebble.NoSync`, we still keep the WAL enabled. Pebble
+		// will fsync the WAL during shutdown and should ensure the db is
+		// recoverable if shutdown correctly.
+		WALBytesPerSync:             cfg.WALBytesPerSync,
+		MemTableStopWritesThreshold: cfg.MemTableStopWritesThreshold,
+		MemTableSize:                cfg.MemTableSize,
+		MaxOpenFiles:                cfg.MaxOpenFiles,
+		MaxConcurrentCompactions:    runtime.NumCPU,
+		Levels:                      make([]pebble.LevelOptions, 7),
```
Comment (joshua-kim) [inline database/pebble/db.go:80]: Wondering if we should configure these levels like geth does instead of leaving these as defaults

File: database/pebble/db.go [hunk 35]
```diff
@@ -0,0 +1,463 @@
...
+		CacheSize:                   units.GiB,
+		BytesPerSync:                units.MiB,
+		WALBytesPerSync:             units.MiB,
+		MemTableStopWritesThreshold: 8,
+		MemTableSize:                16 * units.MiB,
+		MaxOpenFiles:                4 * units.KiB,
+	}
+}
+
+func New(file string, cfg Config, log logging.Logger, _ string, _ prometheus.Registerer) (database.Database, error) {
```
Comment (joshua-kim) [inline database/pebble/db.go:66]: I think we want metrics here similar to what we have for our level db wrapper

**Comment by joshua-kim** [comment]
Moving this back into draft, I think there's a number of things we want to look into (e.g metrics support not present in current revision, want to add some more tests, ...).

File: database/pebble/db.go (hunk 27 above)
Comment (danlaine) [reply database/pebble/db.go]: 🤷 The database.Database tests enforce this

File: database/pebble/db.go [hunk 36]
```diff
@@ -0,0 +1,341 @@
...
+	// https://github.com/ethereum/go-ethereum/blob/release/1.12/ethdb/pebble/pebble.go
+	opts := &pebble.Options{
+		Cache:                       pebble.NewCache(int64(cfg.CacheSize)),
+		BytesPerSync:                cfg.BytesPerSync,
+		Comparer:                    comparer,
+		WALBytesPerSync:             cfg.WALBytesPerSync,
+		MemTableStopWritesThreshold: cfg.MemTableStopWritesThreshold,
+		MemTableSize:                cfg.MemTableSize,
+		MaxOpenFiles:                cfg.MaxOpenFiles,
+		MaxConcurrentCompactions:    runtime.NumCPU, // TODO what should default be?
```
Comment (patrick-ogrady) [inline database/pebble/db.go:85]: You should make this a config that defaults to 1 IMO

File: database/pebble/iterator.go [hunk 37]
```diff
@@ -0,0 +1,159 @@
...
+	}
+
+	if !it.hasNext {
+		return false
+	}
+
+	// Set the next key.
+	it.nextKey = it.iter.Key()
+
+	// Set the next value.
```
Comment (patrick-ogrady) [inline database/pebble/iterator.go:75]: You need to set `it.Iter.First()` somewhere (or at least in my experience)

File: database/pebble/db.go [hunk 38]
```diff
@@ -0,0 +1,341 @@
...
+	ErrInvalidOperation = errors.New("invalid operation")
+
+	defaultCacheSize = 512 * units.MiB
+	DefaultConfig    = Config{
+		CacheSize:                   defaultCacheSize,
+		BytesPerSync:                512 * units.KiB,
+		WALBytesPerSync:             0, // Default to no background syncing.
+		MemTableStopWritesThreshold: 8,
+		MemTableSize:                defaultCacheSize / 4,
+		MaxOpenFiles:                4 * units.KiB,
```
Comment (patrick-ogrady) [inline database/pebble/db.go:50]: This should be the number of files, not a size:

https://github.com/cockroachdb/pebble/blob/1efa535da00668c2b50ac1a3497a2623338db01e/options.go#L786-L790

File: database/pebble/db.go [hunk 39]
```diff
@@ -0,0 +1,341 @@
...
+
+	switch {
+	case db.closed:
+		return database.ErrClosed
+	case comparer.Compare(start, limit) >= 0:
+		// pebble's Compact will no-op & error if start >= limit
+		// according to pebble's comparer.
+		return nil
+	case limit != nil:
+		return updateError(db.pebbleDB.Compact(start, limit, true /* parallelize */))
```
Comment (patrick-ogrady) [inline database/pebble/db.go:206]: I think this should be `false`

File: database/pebble/db.go [hunk 40]
```diff
@@ -0,0 +1,341 @@
...
+	default:
+		return err
+	}
+}
+
+// Returns a key range that covers all keys with the
+// given [prefix].
+// Assumes the Database uses bytes.Compare for key
+// comparison and not a custom comparer.
+func prefixBounds(prefix []byte) *pebble.IterOptions {
```
Comment (patrick-ogrady) [inline database/pebble/db.go:326]: Where did this come from?

File: database/pebble/db.go (hunk 38 above)
Comment (danlaine) [reply database/pebble/db.go:50]: This is equivalent to `4_096`, which is intended, but I'll change it to be declared that way to make it more clear

File: database/pebble/db.go (hunk 39 above)
Comment (danlaine) [reply database/pebble/db.go:206]: Why? It looks like this method is synchronous in either case.

File: database/pebble/db.go (hunk 40 above)
Comment (danlaine) [reply database/pebble/db.go:326]: I'm not sure what you mean. This is how we generate iteration bounds when we use a prefix.

File: database/pebble/db.go (hunk 39 above)
Comment (patrick-ogrady) [reply database/pebble/db.go:206]: If you don't agree, please make it a function arg?

This will spin up a number of goroutines to concurrently compact different ranges...some users may not want this.

Comment (danlaine) [reply database/pebble/db.go]: made this a config

**Comment by danlaine** [comment]
Closed in favor of https://github.com/ava-labs/avalanchego/pull/1999
